File: src/OpenSeries.Lib/Devices/Mice/Aerox5.cs
using HidSharp;
using OpenSeries.Protocols;

namespace OpenSeries.Devices.Mice;

internal sealed class Aerox5Definition : IDeviceDefinition
{
    private static readonly int[] ProductIds =
    [
        0x1850, // Aerox 5
        0x1852, // Aerox 5 Wireless, 2.4 GHz
        0x1854, // Aerox 5 Wireless, wired
        0x185c, // Aerox 5 Wireless Destiny 2 Edition, 2.4 GHz
        0x185e, // Aerox 5 Wireless Destiny 2 Edition, wired
        0x1860, // Aerox 5 Wireless Diablo IV Edition, 2.4 GHz
        0x1862  // Aerox 5 Wireless Diablo IV Edition, wired
    ];

    public string Slug => "aerox-5";

    public bool Matches(HidDevice endpoint)
    {
        if (!ProductIds.Contains(endpoint.ProductID))
        {
            return false;
        }

        try
        {
            return endpoint.GetReportDescriptor().DeviceItems.Any(
                item => item.Usages.ContainsValue(0xffc00001));
        }
        catch
        {
            return endpoint.DevicePath.Contains("&mi_03", StringComparison.OrdinalIgnoreCase);
        }
    }

    public ISteelSeriesDevice Connect(HidDevice endpoint, DeviceIdentity identity) =>
        new Aerox5(endpoint, identity);
}

internal sealed class Aerox5(HidDevice endpoint, DeviceIdentity identity) : IMouseDevice
{
    private static readonly int[] ReceiverProductIds = [0x1852, 0x185c, 0x1860];
    private static readonly int[] WirelessProductIds = [0x1852, 0x1854, 0x185c, 0x185e, 0x1860, 0x1862];
    private static readonly byte[] SkippedSensitivityValues =
    [
        0x08, 0x0f, 0x15, 0x1c, 0x22, 0x24, 0x2b, 0x31, 0x37, 0x3d, 0x43,
        0x49, 0x4f, 0x55, 0x5b, 0x61, 0x67, 0x6d, 0x73, 0x79, 0x7f, 0x85,
        0x8b, 0x91, 0x97, 0x9d, 0xa3, 0xa9, 0xaf, 0xbe, 0xc1, 0xc8, 0xce, 0xd4
    ];

    private const int IoTimeoutMilliseconds = 2_000;
    private const int ReceiverResponseLength = 64;
    private const int CommandDelayMilliseconds = 50;

    private bool IsReceiver => ReceiverProductIds.Contains(ProductId);
    private bool IsWirelessModel => WirelessProductIds.Contains(ProductId);

    public string Id => identity.Id;
    public string Name => ProductId switch
    {
        0x1850 => "SteelSeries Aerox 5",
        0x185c or 0x185e => "SteelSeries Aerox 5 Wireless Destiny 2 Edition",
        0x1860 or 0x1862 => "SteelSeries Aerox 5 Wireless Diablo IV Edition",
        _ => "SteelSeries Aerox 5 Wireless"
    };
    public int ProductId => identity.ProductId;
    public string? SerialNumber => identity.SerialNumber;

    public Features SupportedFeatures =>
        Features.MouseSensitivity |
        Features.PollingRate |
        Features.Illumination |
        (IsWirelessModel ? Features.BatteryStatus | Features.SleepTimer : Features.None);

    public MouseSensitivityInfo SensitivityInfo { get; } = new(100, 18_000, 100, 5);
    public IReadOnlyList<ushort> SupportedPollingRates { get; } = [125, 250, 500, 1000];
    public IReadOnlyList<MouseZone> SupportedIlluminationZones { get; } =
        [MouseZone.Top, MouseZone.Middle, MouseZone.Bottom];

    public void SetSensitivity(IReadOnlyList<ushort> dpiPresets)
    {
        ArgumentNullException.ThrowIfNull(dpiPresets);
        if (dpiPresets.Count is < 1 or > 5)
        {
            throw new ArgumentException("Between one and five DPI presets are required.", nameof(dpiPresets));
        }

        var command = new byte[3 + dpiPresets.Count];
        command[0] = 0x2d;
        command[1] = (byte)dpiPresets.Count;
        command[2] = 0;
        for (int index = 0; index < dpiPresets.Count; index++)
        {
            ushort dpi = dpiPresets[index];
            if (dpi is < 100 or > 18_000 || dpi % 100 != 0)
            {
                throw new ArgumentOutOfRangeException(
                    nameof(dpiPresets), $"DPI preset {index + 1} must be from 100 to 18000 in steps of 100.");
            }

            command[index + 3] = EncodeSensitivity(dpi);
        }

        SendAndSave(command);
    }

    public void SetPollingRate(ushort pollingRate)
    {
        byte encoded = pollingRate switch
        {
            125 => IsWirelessModel ? (byte)0x03 : (byte)0x04,
            250 => IsWirelessModel ? (byte)0x02 : (byte)0x03,
            500 => IsWirelessModel ? (byte)0x01 : (byte)0x02,
            1000 => IsWirelessModel ? (byte)0x00 : (byte)0x01,
            _ => throw new ArgumentOutOfRangeException(
                nameof(pollingRate), "Polling rate must be 125, 250, 500, or 1000 Hz.")
        };

        SendAndSave([0x2b, encoded]);
    }

    public void SetIllumination(MouseZone zone, RgbColor color)
    {
        ArgumentNullException.ThrowIfNull(color);
        if (zone is not (MouseZone.Top or MouseZone.Middle or MouseZone.Bottom))
        {
            throw new ArgumentOutOfRangeException(
                nameof(zone), "Aerox 5 lighting zone must be top, middle, or bottom.");
        }

        byte[] prefix = IsWirelessModel
            ? [0x21, 0x01, (byte)zone]
            : zone switch
            {
                MouseZone.Top => [0x21, 0x01],
                MouseZone.Middle => [0x21, 0x02, 0x00, 0x00, 0x00],
                MouseZone.Bottom => [0x21, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00],
                _ => throw new ArgumentOutOfRangeException(nameof(zone))
            };

        SendAndSave([.. prefix, color.Red, color.Green, color.Blue]);
    }

    public void SetSleepTimer(byte minutes)
    {
        if (!IsWirelessModel)
        {
            throw new NotSupportedException("Sleep timer is only available on Aerox 5 Wireless models.");
        }
        if (minutes > 20)
        {
            throw new ArgumentOutOfRangeException(nameof(minutes), "Sleep timer must be between 0 and 20 minutes.");
        }

        int milliseconds = minutes * 60_000;
        SendAndSave(
        [
            0x29,
            (byte)milliseconds,
            (byte)(milliseconds >> 8),
            (byte)(milliseconds >> 16)
        ]);
    }

    public BatteryInfo GetBattery()
    {
        if (!IsWirelessModel)
        {
            throw new NotSupportedException("Battery status is only available on Aerox 5 Wireless models.");
        }

        byte[] response = SendCommand([0x92], true);
        byte responseCommand = IsReceiver ? (byte)0xd2 : (byte)0x92;
        int offset = response.Length >= 3 && response[0] == 0x00 && response[1] == responseCommand ? 1 : 0;
        if (response.Length - offset < 2)
        {
            throw new InvalidDataException($"Device returned a short battery response ({response.Length} bytes).");
        }

        byte raw = response[offset + 1];
        int batteryStep = raw & 0x7f;
        if (batteryStep is < 1 or > 21)
        {
            throw new InvalidOperationException("mouse is disconnected or returned an invalid battery level");
        }

        ushort level = (ushort)((batteryStep - 1) * 5);
        BatteryStatus status = (raw & 0x80) != 0
            ? BatteryStatus.Charging
            : level == 100 ? BatteryStatus.Charged : BatteryStatus.Discharging;
        return new BatteryInfo(level, status, response);
    }

    private void SendAndSave(ReadOnlySpan<byte> command)
    {
        SendCommand(command, IsReceiver);
        Thread.Sleep(CommandDelayMilliseconds);
        SendCommand([0x11, 0x00], IsReceiver);
    }

    private byte[] SendCommand(ReadOnlySpan<byte> command, bool readResponse)
    {
        using HidStream stream = endpoint.Open();
        stream.ReadTimeout = IoTimeoutMilliseconds;
        stream.WriteTimeout = IoTimeoutMilliseconds;

        int reportLength = endpoint.GetMaxOutputReportLength();
        if (reportLength < command.Length + 1)
        {
            throw new InvalidDataException($"Output report length {reportLength} cannot carry this command.");
        }

        byte[] report = new byte[reportLength];
        command.CopyTo(report.AsSpan(1));
        if (IsReceiver)
        {
            report[1] |= 0x40;
        }
        stream.Write(report);

        if (!readResponse)
        {
            return [];
        }

        var response = new byte[Math.Max(ReceiverResponseLength, endpoint.GetMaxInputReportLength())];
        int bytesRead = stream.Read(response);
        return response[..bytesRead];
    }

    private static byte EncodeSensitivity(ushort dpi)
    {
        if (dpi == 100)
        {
            return 0;
        }

        int encoded = dpi / 100;
        foreach (byte skipped in SkippedSensitivityValues)
        {
            if (encoded < skipped)
            {
                break;
            }
            encoded++;
        }
        return checked((byte)encoded);
    }
}

File: src/devices/mice/aerox_5.rs
use super::*;

const AEROX_5_ZONES: [MouseZone; 3] = [MouseZone::Top, MouseZone::Middle, MouseZone::Bottom];
const SKIPPED_SENSITIVITY: [u8; 35] = [
    0x08, 0x0f, 0x15, 0x1c, 0x22, 0x24, 0x2b, 0x31, 0x37, 0x3d, 0x43, 0x49, 0x4f, 0x55, 0x5b, 0x61,
    0x67, 0x6d, 0x73, 0x79, 0x7f, 0x85, 0x8b, 0x91, 0x97, 0x9d, 0xa3, 0xa9, 0xaf, 0xbe, 0xc1, 0xc8,
    0xce, 0xd4, 0xff,
];

pub(crate) static DEFINITION: DeviceDefinition = DeviceDefinition {
    slug: "aerox-5",
    product_ids: &[0x1850, 0x1852, 0x1854, 0x185c, 0x185e, 0x1860, 0x1862],
    rule: MatchRule::VendorControl,
    connect: |identity, transport| {
        Device::Mouse(Mouse::new(Box::new(Aerox5::new(identity, transport))))
    },
};

struct Aerox5 {
    base: DeviceContext,
}

impl Aerox5 {
    fn new(identity: Identity, transport: HidTransport) -> Self {
        Self {
            base: DeviceContext::new(identity, transport),
        }
    }

    fn receiver(&self) -> bool {
        [0x1852, 0x185c, 0x1860].contains(&self.base.identity.product_id)
    }

    fn wireless(&self) -> bool {
        [0x1852, 0x1854, 0x185c, 0x185e, 0x1860, 0x1862].contains(&self.base.identity.product_id)
    }

    fn send(&mut self, command: &[u8], read: bool) -> Result<Vec<u8>> {
        let mut framed = command.to_vec();
        if self.receiver() {
            framed[0] |= 0x40;
        }
        if read {
            self.base
                .transport
                .write_output_and_read(&framed, 64, 1, 0, None)
        } else {
            self.base.transport.write_output(&framed, 1, 0)?;
            Ok(Vec::new())
        }
    }

    fn send_and_save(&mut self, command: &[u8]) -> Result<()> {
        let receiver = self.receiver();
        self.send(command, receiver)?;
        thread::sleep(Duration::from_millis(50));
        self.send(&[0x11, 0], receiver)?;
        Ok(())
    }

    fn encode_sensitivity(dpi: u16, index: usize) -> Result<u8> {
        if !(100..=18_000).contains(&dpi) || !dpi.is_multiple_of(100) {
            return Err(invalid(format!(
                "DPI preset {} must be from 100 to 18000 in steps of 100.",
                index + 1
            )));
        }
        if dpi == 100 {
            return Ok(0);
        }
        let mut encoded = usize::from(dpi / 100);
        for skipped in SKIPPED_SENSITIVITY {
            if encoded < usize::from(skipped) {
                break;
            }
            encoded += 1;
        }
        u8::try_from(encoded).map_err(|_| invalid("DPI encoding overflow."))
    }

    fn battery(&mut self) -> Result<BatteryInfo> {
        if !self.wireless() {
            return Err(unsupported(
                "Battery status is only available on Aerox 5 Wireless models.",
            ));
        }
        let receiver = self.receiver();
        let response = self.send(&[0x92], true)?;
        let command = if receiver { 0xd2 } else { 0x92 };
        let offset = usize::from(response.len() >= 3 && response[0] == 0 && response[1] == command);
        if response.len().saturating_sub(offset) < 2 {
            return Err(OpenSeriesError::Protocol(format!(
                "Device returned a short battery response ({} bytes).",
                response.len()
            )));
        }
        let raw = response[offset + 1];
        let step = raw & 0x7f;
        if !(1..=21).contains(&step) {
            return Ok(BatteryInfo {
                level_percentage: 0,
                status: BatteryStatus::Disconnected,
            });
        }
        let level = u16::from(step - 1) * 5;
        let status = if raw & 0x80 != 0 {
            BatteryStatus::Charging
        } else if level == 100 {
            BatteryStatus::Charged
        } else {
            BatteryStatus::Discharging
        };
        Ok(BatteryInfo {
            level_percentage: level,
            status,
        })
    }
}

impl DeviceProtocol for Aerox5 {
    fn id(&self) -> &str {
        &self.base.identity.id
    }

    fn name(&self) -> &str {
        match self.base.identity.product_id {
            0x1850 => "SteelSeries Aerox 5",
            0x185c | 0x185e => "SteelSeries Aerox 5 Wireless Destiny 2 Edition",
            0x1860 | 0x1862 => "SteelSeries Aerox 5 Wireless Diablo IV Edition",
            _ => "SteelSeries Aerox 5 Wireless",
        }
    }

    fn product_id(&self) -> u16 {
        self.base.identity.product_id
    }

    fn supported_features(&self) -> Capabilities {
        Capabilities::MOUSE_SENSITIVITY
            | Capabilities::POLLING_RATE
            | Capabilities::ILLUMINATION
            | if self.wireless() {
                Capabilities::BATTERY_STATUS | Capabilities::SLEEP_TIMER
            } else {
                Capabilities::empty()
            }
    }
}

impl MouseProtocol for Aerox5 {
    fn sensitivity_info(&self) -> Result<MouseSensitivityInfo> {
        Ok(MouseSensitivityInfo {
            minimum: 100,
            maximum: 18_000,
            step: 100,
            maximum_preset_count: 5,
        })
    }

    fn supported_polling_rates(&self) -> Result<&'static [u16]> {
        Ok(&POLLING_RATES)
    }

    fn supported_illumination_zones(&self) -> Result<&'static [MouseZone]> {
        Ok(&AEROX_5_ZONES)
    }

    fn set_sensitivity(&mut self, values: &[u16]) -> Result<()> {
        if !(1..=5).contains(&values.len()) {
            return Err(invalid("Between one and five DPI presets are required."));
        }
        let mut command = vec![
            0x2d,
            values.len() as u8,
            if self.wireless() { 0 } else { 1 },
        ];
        for (index, dpi) in values.iter().copied().enumerate() {
            command.push(Self::encode_sensitivity(dpi, index)?);
        }
        self.send_and_save(&command)
    }

    fn set_polling_rate(&mut self, rate: u16) -> Result<()> {
        let encoded = match (rate, self.wireless()) {
            (125, true) => 3,
            (250, true) => 2,
            (500, true) => 1,
            (1000, true) => 0,
            (125, false) => 4,
            (250, false) => 3,
            (500, false) => 2,
            (1000, false) => 1,
            _ => return Err(invalid("Polling rate must be 125, 250, 500, or 1000 Hz.")),
        };
        self.send_and_save(&[0x2b, encoded])
    }

    fn set_illumination(
        &mut self,
        zone: MouseZone,
        color: RgbColor,
        persistence: Persistence,
    ) -> Result<()> {
        if !AEROX_5_ZONES.contains(&zone) {
            return Err(invalid(
                "Aerox 5 lighting zone must be top, middle, or bottom.",
            ));
        }
        let mut command = if self.wireless() {
            vec![0x21, 1, zone as u8]
        } else {
            match zone {
                MouseZone::Top => vec![0x21, 1],
                MouseZone::Middle => vec![0x21, 2, 0, 0, 0],
                MouseZone::Bottom => vec![0x21, 4, 0, 0, 0, 0, 0, 0],
                _ => unreachable!(),
            }
        };
        command.extend([color.red, color.green, color.blue]);
        if persistence == Persistence::Save {
            self.send_and_save(&command)
        } else {
            self.send(&command, self.receiver()).map(drop)
        }
    }

    fn set_sleep_timer(&mut self, minutes: u8) -> Result<()> {
        if !self.wireless() {
            return Err(unsupported(
                "Sleep timer is only available on Aerox 5 Wireless models.",
            ));
        }
        if minutes > 20 {
            return Err(invalid("Sleep timer must be between 0 and 20 minutes."));
        }
        let ms = u32::from(minutes) * 60_000;
        let bytes = ms.to_le_bytes();
        self.send_and_save(&[0x29, bytes[0], bytes[1], bytes[2]])
    }

    fn get_battery(&mut self) -> Result<BatteryInfo> {
        self.battery()
    }
}
